### DISCUSSION TITLE
Sharding a single data product

### DISCUSSION THREAD

**Discussion by matko** [terminusdb/terminusdb#526] [Ideas]
Since @GavinMendelGleason just wrote #518 , I feel like I should add to it with other forms of sharding. Domain sharding is great, and we should have it, but this sort of sharding really only works when the data products move entirely independently. There's another form of sharding that people want, which is sharding within a single data product, where the different shards are not independent from each other.

What's the difference? In domain sharding, we shard different kinds of types into different graphs. The sharding criteria is the domain of these types. Within a single domain, the sharding criteria has to be something inside of these types themselves.

## the point of sharding within a single data product
So what is the point of this sort of sharding? There's a couple.

### Breaking up transaction domains
One of the biggest reasons for doing this is that it'll allow concurrent write transactions, as long as these transactions happen in a different shard. Each shard would effectively be its own transaction domain.

### Smaller runtime memory footprint
When querying, the only graphs that need to be loaded are graphs actually looked at. For very large graphs, this can significantly reduce the memory footprint.

### Retain a shared history
This is more of a comparison with domain sharding. Sharding within a single domain retains a shared history, conceptually anyway. By guaranteeing that shards cannot contain any sort of overlapping data, the history of the sharded domain can be reconstructed by merging the individual histories of each shard. We could easily convert a sharded database back to one without shards, or one with more shards, by splitting the history accordingly.

### Federated queries
A bit more future talk, but it could be possible some time to actually have different shards loaded on different machines, and use a federated protocol to perform a large query. This'd allow us to do queries on data sets that are too large to fit on any one machine, thus scaling to petabytes and beyond.

## When this is desirable
Sharding would be very useful in situations where you have a lot of data that is receiving a lot of writes, but the individual writes tend to be isolated to just a small part of the graph. This is for example the case with our own system database, which holds user, organization and database data. Individual queries tend to only look at one organization and its associated database, and only some users. Right now, concurrent writes to the system graph are problematic. If we broke up the transaction domain with shards, these problems would (largely) go away.

### Sharding criteria
With our object model we have a pretty natural sharding strategy available to us. We can simply shard on document ids. Any subdocuments should live in the same shard as the document that points at them. ValueHash subdocuments can be replicated in all shards that use them.

We need some function that takes an object iri and turns it into a shard number. Ideally, this is something that the user is able to set up themselves. Shard numbers can be some arbitrary large number (though maybe we should limit it to a u32 or a u64), which we can then take a modulo of to determine the actual shard this object should live in.

With such a sharding strategy we can be sure that no shard contains duplicate data (except potentially duplicate ValueHash subdocuments, which is ok).

### Retrieval
When retrieving documents by ids, we can use the same procedure to determine in what shard to look. Then, all subdocuments should live in the same graph. Similarly for updating documents by id, we know exactly where to store the document and all its subdocuments.

For more complex queries, such as "find all people with 'mark' as first name", we are still required to go over all shards, but having found a match in a shard, we know that any directly connected object (the owner document and the contained subdocument) will live in the same shard.

### History and metadata graphs
Each shard would have its own history. In other words, each shard would have its own 'shard graph', that contains the commit information relevant to that shard, and points to individual shard layers containing the actual data. To properly split up transaction domains, these shard graphs should be directly pointed at by a store label, so that they can be updated independently. We still ought to point at them through the existing metadata graph structures though.

In my opinion, shards should live on the branch/named graph level. When sharding a named graph, we ought to create a special commit on them which contains pointers to all the shards, as well as the instruction on the specific sharding strategy. After that, all further updates are done in the shard graphs, until such a time that the shards are updated.
When updating the sharding strategy or the shard count, we can merge back these individual histories into the commit graph, and then create a new shard commit on top of that with the new strategy. This way we always maintain the history properly.

### Complex updates
When updates need to involve multiple shards at once, we need to do something that's being planned for a while now: multi-domain transactions. Plans for this have been drawn up but sort of assume big transaction domains delineated by a metadata graph. This would no longer be the case with these shards.
Instead, each shard needs its own metadata graph, where transactions can live. The total graph structure then becomes

metadata -> commit -> shard meta -> shard commit -> data. 

This may seem pretty excessive, but the extra overhead may be worth it for the cases where one wants sharding.

## What we need before we could implement this
- Multi-domain transactions
- (Ideally) historyless graphs (for the new metadata layer)

**Comment by matko** [comment]
Probably, we are only interested in sharding the instance graph. the schema graph is small enough to load fully all the time, and we probably need the whole thing all the time anyway.

**Comment by GavinMendelGleason** [comment]
This is fantastic!  In terms of the "what we need", do you have any thoughts on implementation of historyless graphs?

**Comment by matko** [comment]
> This is fantastic! In terms of the "what we need", do you have any thoughts on implementation of historyless graphs?

I've really not given that much thought. But it's been under consideration for our various (often small) graph types which see a lot of updates but for which we don't really care about the history at all, such as all the metadata graphs. Assuming we have such a thing, it makes sense to use that to keep track of any ongoing transactions in shards too.

I suspect any sort of persistent hash structure would do, and we should just evaluate a bunch and see what works. If we put a compatible api on it, terminus-server doesn't even have to notice the difference between different graph types.
